Recruit-Training/Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +23,58 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return amphour_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    return (current_ < 0);
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return QTime(0, minutes_, seconds_, miliseconds_);
 }
 
+int BatteryStateOfChargeService::getHours() const
+{
+    return hours_;
+}
+
+
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+    double currentOld_ = current_;
+    current_ = batteryData.current;
+
+    QTime timeOld_ = time;
+    time = batteryData.time;
+
+    double averageAmphour_ = (current_ + currentOld_) / 2;
+
+    //timeOld_ would be null for the first entry of data
+    if (!timeOld_.isNull())
+    {
+        amphour_ += averageAmphour_ * timeOld_.msecsTo(time) / HOURS_TO_MILISECONDS;
+    }
+
+    //Calculation of the time until charged/depleted
+    if (isCharging())
+    {
+        totalTimeHours_ =  abs(amphour_ / current_);
+    }
+    else
+    {
+        totalTimeHours_ =  (BATTERY_AMP_HOUR_CAPACITY - amphour_) / current_;
+    }
+
+    //Rounding the total time (in hours) down
+    hours_ = totalTimeHours_;
```

Comment:
I wouldn't know that you were rounding it down without your comment, if I didn't know `hours_` was an int & `totalTimeHours_` was a double. I'd say either to rename these two variables so the type is clear, or add a explicit double to int cast.

Better yet, I think the `floor` function might be the most explicit way.